var/spack/repos/builtin/packages/python/package.py

```diff
@@ -686,87 +687,58 @@ def print_string(self, string):
-        cmd += self.print_string('get_python_inc({0})'.format(args))
-
-        return self.command('-c', cmd, output=str).strip()
-
-    def get_python_lib(self, plat_specific=False, standard_lib=False, prefix=None):
-        """Return the directory for either the general or platform-dependent
-        library installation. Wrapper around ``distutils.sysconfig.get_python_lib()``.
+    @property
+    def config_vars(self):
+        """Return a set of variable definitions associated with a Python installation.
 
-        Parameters:
-            plat_specific (bool): if true, the platform-dependent library directory
-                is returned, else the platform-independent directory is returned
-            standard_lib (bool): if true, the directory for the standard library is
-                returned rather than the directory for the installation of
-                third-party extensions
-            prefix (str): prefix to use instead of ``distutils.sysconfig.PREFIX``
+        Wrapper around various ``distutils.sysconfig`` functions.
 
         Returns:
-            str: library installation directory
+            dict: variable definitions
         """
-        # Wrap strings in quotes
-        if prefix is not None:
-            prefix = '"{0}"'.format(prefix)
+        # TODO: distutils is deprecated in Python 3.10 and will be removed in
+        # Python 3.12, find a different way to access this information.
+        # Also, calling the python executable disallows us from cross-compiling,
+        # so we want to try to avoid that if possible.
+        cmd = """
+import json
+from distutils.sysconfig import (
+    get_config_vars,
+    get_config_h_filename,
+    get_makefile_filename,
+    get_python_inc,
+    get_python_lib,
+)
+
+config = get_config_vars()
+config['config_h_filename'] = get_config_h_filename()
+config['makefile_filename'] = get_makefile_filename()
+config['python_inc'] = get_python_inc(prefix='')
+config['python_lib'] = {}
+
+for plat_specific in [True, False]:
+    config['python_lib'][plat_specific] = {}
+    for standard_lib in [True, False]:
+        config['python_lib'][plat_specific][standard_lib] = get_python_lib(
+            plat_specific, standard_lib, prefix=''
+        )
+
+%s
+""" % self.print_string("json.dumps(config)")
 
-        args = 'plat_specific={0}, standard_lib={1}, prefix={2}'.format(
-            plat_specific, standard_lib, prefix)
-
-        cmd = 'from distutils.sysconfig import get_python_lib; '
-        cmd += self.print_string('get_python_lib({0})'.format(args))
-
-        return self.command('-c', cmd, output=str).strip()
+        dag_hash = self.spec.dag_hash()
+        if dag_hash not in self._config_vars:
+            try:
+                config = json.loads(self.command('-c', cmd, output=str))
+            except (ProcessError, RuntimeError):
+                config = {}
+            self._config_vars[dag_hash] = config
+        return self._config_vars[dag_hash]
```

Comment:
I'm wondering if this can be simplified and you could just cache a single config vars dictionary on the package instance.  I think that will get you a once-per-dag-hash invocation but:

1. won't be sensitive to the hash (so when we go to full hash this won't need to change), and
2. makes the logic a bit simpler here.

Can you try it and see if it's still only called once?

Reply:
In all of the environments I've tried it's only called once, so dag_hash isn't needed. I'm not sure why @alalazo added it originally, maybe he can comment. My understanding was that for now only a single instance of a package can appear in the DAG. Could it be for environments when `concretization: separate`?

Reply:
> I'm not sure why @alalazo added it originally, maybe he can comment.

This originally was for the `home` variable, and if I understand correctly here it's being generalized. The cache is at the class level, not at the instance level, so that in a context with multiple Python packages the instance will compute the value once per dag hash (which is the granularity we have in the store)